pkg/cmd/target_history.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/gardener/gardenctl/pkg/internal/history"
+	"github.com/spf13/cobra"
+)
+
+//NewHistoryCmd use for list/search targting history
+func NewHistoryCmd() *cobra.Command {
+	return &cobra.Command{
+		Use:          "x",
+		Short:        "List/Search targeting history, e.g. \"gardenctl x\"",
+		SilenceUsage: true,
+		Long:         `List Targeting Hisotry`,
+		Run: func(cmd *cobra.Command, args []string) {
+			h := history.Settings(pathHistory, os.Args[0])
+			if len(h.Load().Items) <= 0 {
+				fmt.Println("No Target History results")
+				os.Exit(0)
+			}
+			item := h.Load().Reverse().Select().Items
+			history.Execute(h.Binary, item)
```

Comment:
```go
			h := history.Settings(pathHistory, os.Args[0]) // no need to pass the binary name here as it's only used a few lines down below
	                 // [...]
			history.Execute(h.Binary, item)
```

Reply:
Why did you choose to execute the binary with the previous arguments and let gardenctl figure out the target stack once again. 

1. E.g. assume you target `g --garden dev --seed seed-dev --namespace shoot--someproject--someshoot`
Target is:
```yaml
garden: dev
seed: seed-dev
namespace: shoot--someproject--someshoot
```
2. then you target `g target garden canary`
Target is:
```yaml
garden: canary
```
3. then you target the project `g target project myproject`
Target is:
```yaml
garden: canary
project: myproject
```
4. then you target the shoot `g target shoot myshoot`
Target is:
```yaml
garden: canary
project: myproject
shoot: myshoot
```
5. Then you open the history and select the first entry which will bring you to the dev garden. 
Target is again:
```yaml
garden: dev
seed: seed-dev
namespace: shoot--someproject--someshoot
```
6. Then you select 4. (`g target shoot myshoot`) but this does not work and results in an error `Error: no match for myshoot` as you only pass the arguments `target shoot myshoot` to gardenctl but meanwhile on the target stack is a different garden cluster (`dev` in this cae) and a different project / seed selected.



You already seem to store the target into the `history` file, like
```json
{
  "Cmd":"gardenctl target --server https://example.com --project myproject --shoot myshoot",
  "garden":"dev",
  "project":"myproject",
  "shoot":"myshoot"
}
```
Isn't it easier to just build the target from these values and not by executing gardenctl with the arguments?

Reply:
> ```go
> 			h := history.Settings(pathHistory, os.Args[0]) // no need to pass the binary name here as it's only used a few lines down below
> 	                 // [...]
> 			history.Execute(h.Binary, item)
> ```


after initial `h := history.Settings(pathHistory, os.Args[0])`, I can `use h.Binary`. Could you elaborate more here?  Do you mean  remove `os.Args[0]` then just simple hardcode in `history.Execute("gardenctl", item)` then remove unuse code eg `history.Settings`

which looks like below?
```
const (Binary = "gardenctl")
h := history.SetPath(pathHistory)
//[...]
history.Execute(Binary, item) or just  history.Execute("gardenctl", item)

Reply:
> Why did you choose to execute the binary with the previous arguments and let gardenctl figure out the target stack once again.

My intention is `history` only persists in targeting history without any new target logic for history feature.  the target part still uses by `gardenctl` to figure out the target stack. but I got your point. I will take a look whether I can use `history` file to feed/inject `target` without adding extra code in `target.go` as it already heavy 

{
  "Cmd":"gardenctl target --server https://example.com --project myproject --shoot myshoot",
  "garden":"dev",
  "project":"myproject",
  "shoot":"myshoot"
}

Reply:
done  in d87da4c442c8b11e4eedc225005963891f076e28 change logic load from `history` file then writing Target directly to local `target` file when switching from the history items.

Reply:
reopen the conversation, 
find one bug, history file order break for in scenario target control plane,  due to json Marshal alphabet order from the history file.  
as a result, rewrite target in the wrong order as well. which causes kubeconfig path not able to find. 
will take a look whether able to fix it

Reply:
change the logic to fixed the bug, due to golang `map` unorder feature. it will make `for kind, name := range m` random order which caused the wrong target to create sometimes.

Reply:
another issue found due `Namespace` switch by `gardenctl target` thought `kubectx` , however history only rewrite the lcoal `target` file which caused current context does not to switch `Namespace`

```
Garden:
KUBECONFIG=xxx
Seed:
KUBECONFIG=xxx
Namespace:
Set namespace to shoot--ixxxxxfor current context shoot-garden-gcp-eu1
Context "shoot-garden-gcp-eu1" modified.
```

Reply:
done, in e9aec0ef73f2f786ca9becf593d0dddef3fe7a7e, fix `Namespace` issue